Docs/developer_guide/parameter_nodes.md

```diff
@@ -0,0 +1,366 @@
+@parameterNodeSerializer
+class CustomClassSerializer(Serializer):
+  @staticmethod
+  def canSerialize(type_) -> bool:
+    """
+    Whether the serializer can serialize the given type if it is properly instantiated.
+    """
+    return type_ == CustomClass
+
+  @staticmethod
+  def create(type_):
+    """
+    Creates a new serializer object based on the given type. If this class does not support the given type,
+    None is returned.
+
+    It is common for the returned type to actually be a ValidatedSerializer wrapping this serializer that implements
+    any default validators (NotNone and IsInstance are common).
+    """
+    if CustomClassSerializer.canSerialize(type_):
+      # in our example, lets say that we don't allow None. We will use NotNone() to enforce this
+      return ValidatedSerializer(CustomClassSerializer(), [NotNone(), IsInstance(CustomClass)])
+    return None
+
+  def default(self):
+    """
+    The default value to use if another default is not specified.
+    """
+    return CustomClass(0, 0, 0)
+
+  def isIn(self, parameterNode: slicer.vtkMRMLScriptedModuleNode, name: str) -> bool:
+    """
+    Whether the parameterNode contains a parameter of the given name.
+    Note that most implementations can just use parameterNode.HasParameter(name).
+    """
+    return parameterNode.HasParameter(name)
+  def write(self, parameterNode: slicer.vtkMRMLScriptedModuleNode, name: str, value) -> None:
+    """
+    Writes the value to the parameterNode under the given name.
+    Note: It is acceptable to mangle the name as long the same name can be used for reading.
+    For example the built-in ListSerializer does this.
+    """
+    parameterNode.SetParameter(name, f"{value.x},{value.y},{value.z}")
+  def read(self, parameterNode: slicer.vtkMRMLScriptedModuleNode, name: str):
+    """
+    Reads and returns the value with the given name from the parameterNode.
+    """
+    val = parameterNode.GetParameter(name)
+    vals = val.split(',')
+    return CustomClass(int(vals[0]), int(vals[1]), int(vals[2]))
+  def remove(self, parameterNode: slicer.vtkMRMLScriptedModuleNode, name: str) -> None:
+    """
+    Removes the value of the given name from the parameterNode.
+    """
+    parameterNode.UnsetParameter(name)
+
+@parameterNodeWrapper
+class CustomClassParameterNode(object):
+  # can now use CustomClass like any other type for building parameterNodeWrappers
+  custom: Annotated[CustomClass, Default(CustomClass(1,2,3))]
+  listOfCustom: list[CustomClass]
+
+```
+
+### Caching
+
+The `vtkMRMLScriptedModuleNode` stores its values as strings, but converting from string to another
+type can be slow. Caching is used to speed up repeated reads under the hood. If the value hasn't been written to the _parameter node wrapper_ then it will use the cached value. Note this means if the underlying `vtkMRMLScriptedModuleNode` parameter node changes not through the wrapper, the cache will not be updated and the cached value will be wrong.
```

Comment:
```suggestion
type can be slow. Caching is used to reduce repeated costly reads under the hood. If a value hasn't been written through the _parameter node wrapper_ since the last read, then it will use the cached value. Note this means if the underlying `vtkMRMLScriptedModuleNode` parameter node changes outside of the wrapper, the cache will not be updated and the cached value will be wrong.
```

Reply:
Updated.